src/main/java/com/redhat/devtools/lsp4ij/usages/LSPFindUsagesHandler.java

```diff
@@ -0,0 +1,67 @@
+package com.redhat.devtools.lsp4ij.usages;
+
+import com.intellij.find.findUsages.FindUsagesHandler;
+import com.intellij.find.findUsages.FindUsagesOptions;
+import com.intellij.openapi.application.ReadAction;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.psi.PsiElement;
+import com.intellij.usageView.UsageInfo;
+import com.intellij.util.Processor;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import org.eclipse.lsp4j.Position;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class LSPFindUsagesHandler extends FindUsagesHandler {
+    public LSPFindUsagesHandler(@NotNull PsiElement psiElement) {
+        super(psiElement);
+    }
+
+    @Override
+    public boolean processElementUsages(@NotNull PsiElement element, @NotNull Processor<? super UsageInfo> processor, @NotNull FindUsagesOptions options) {
+        LSPUsageTriggeredPsiElement elt = (LSPUsageTriggeredPsiElement) element;
+        if (elt.getLSPReferences() != null) {
+            elt.getLSPReferences()
+                    .forEach(ref -> {
+                        processor.process(ReadAction.compute(() -> new UsageInfo(LSPUsagesManager.toPsiElement(ref, element.getProject()))));
+                    });
+        } else {
+            Document document = LSPIJUtils.getDocument(element.getContainingFile().getVirtualFile());
+            Position position = LSPIJUtils.toPosition(element.getTextRange().getStartOffset(), document);
+
+            LSPUsageSupport.LSPUsageSupportParams params = new LSPUsageSupport.LSPUsageSupportParams(position, processor);
+            LSPUsageSupport usageSupport = new LSPUsageSupport(element.getContainingFile());
+            CompletableFuture<List<LSPUsagePsiElement>> usagesFuture = usageSupport.getFeatureData(params);
+            waitFor(usagesFuture);
+            if (usagesFuture.isDone()) {
+                List<LSPUsagePsiElement> usages = usagesFuture.getNow(Collections.emptyList());
+                for (LSPUsagePsiElement usage : usages ) {
+                    processor.process( ReadAction.compute(() ->new UsageInfo(usage)));
+                }
+            }
+        }
+        return true;
+    }
+
+    private static void waitFor(CompletableFuture<?> future) {
+        while (!future.isDone()) {
+            try {
+                future.get(25, TimeUnit.MILLISECONDS);
+                ProgressManager.checkCanceled();
+            } catch (ExecutionException e) {
+
+            } catch (InterruptedException e) {
+                Thread.interrupted();
+            } catch (TimeoutException e) {
+
+            }

```

Comment:
```java
try {
        future.get(25, TimeUnit.MILLISECONDS); //no need for while loop, as get(timeout, unit) already blocks until the future completes or the timeout is reached.
        ProgressManager.checkCanceled();
    } catch (ExecutionException | TimeoutException ignore) {
    } catch (InterruptedException e) {
        Thread.interrupted();
    }
```

Reply:
fixed, I move this method in CompletableFutures class as utility method since this code is used too in LSPCodeLensProvider.